### PR TITLE
convert type-variable to typescript

### PR DESCRIPTION
**Pull request by lindner**
Addresses #1765

### CODE REVIEW

File: runtime/ts/type.ts [hunk 1]
```diff
@@ -64,7 +64,8 @@ export class Type {
     }
     if (tag === 'Variable') {
       if (!(data instanceof TypeVariable)) {
-        data = new TypeVariable(data.name, data.constraint);
+        // TODO determine if 3rd param should be null?
```
Comment (lindner) [inline runtime/ts/type.ts]: any advice for this?

File: runtime/ts/type-variable.ts [hunk 2]
```diff
@@ -7,13 +7,18 @@
...
 export class TypeVariable {
-  constructor(name, canWriteSuperset, canReadSubset) {
-    assert(typeof name == 'string');
+  name: string;
+  _canWriteSuperset: Type|null;
+  _canReadSubset: Type|null;
+  _resolution: Type|null;
+
+  constructor(name: string, canWriteSuperset: Type|null, canReadSubset: Type|null) {
+    assert(typeof name === 'string');
```
Comment (mykmartin) [inline runtime/ts/type-variable.ts]: These asserts seem a bit redundant; can we remove them now?

File: runtime/ts/type.ts (hunk 1 above)
Comment (mykmartin) [reply runtime/ts/type.ts]: @shans should be able to help.

File: runtime/ts/type-variable.ts (hunk 2 above)
Comment (lindner) [reply runtime/ts/type-variable.ts]: done

File: runtime/ts/type.ts (hunk 1 above)
Comment (lindner) [reply runtime/ts/type.ts]: Okay, added as a reviewer.

Comment (shans) [reply runtime/ts/type.ts]: Yeah, I think that the 3rd param should be null here; but it's not immediately clear why, so would you mind adding a comment at the same time:
```
// type constraints ("~a with EntityName") should be considered minimum requirements
// for the type, so are fed in as 'canWriteSuperset' (i.e. low-watermark) constraints.
```

That's still a bit cryptic sorry :/ but at least it provides a hint of what's going on.

Comment (lindner) [reply runtime/ts/type.ts]: thanks, done.

**Pull request merged**